File: CPPBlackJack/src/cardpile.cpp
#include "cardpile.h"

#include <chrono>

uint64_t state =
    std::chrono::high_resolution_clock::now().time_since_epoch().count();

// From https://github.com/lemire/testingRNG
uint64_t wyrand() {
  state += UINT64_C(0xa0761d6478bd642f);
  __uint128_t t = (__uint128_t)state * (state ^ UINT64_C(0xe7037ed1a0b428db));
  return (t >> 64) ^ t;
}

// use nearly divisionless technique found here
// https://github.com/lemire/FastShuffleExperiments
uint64_t rand_range(uint64_t s) {
  uint64_t x = wyrand();
  __uint128_t m = (__uint128_t)x * (__uint128_t)s;
  uint64_t l = (uint64_t)m;
  if (l < s) {
    uint64_t t = -s % s;
    while (l < t) {
      x = wyrand();
      m = (__uint128_t)x * (__uint128_t)s;
      l = (uint64_t)m;
    }
  }
  return m >> 64;
}

CardPile::CardPile(const int num_of_decks) {
  for (auto x = 0; x < num_of_decks; x++) {
    Deck temp_deck;
    m_cards_.insert(m_cards_.end(), temp_deck.m_cards_.begin(),
                    temp_deck.m_cards_.end());
  }
  m_original_cards_ = m_cards_;
}

void CardPile::refresh() { m_cards_ = m_original_cards_; }

std::string CardPile::print() {
  std::string output;
  for (auto& i : m_cards_) {
    output += i->print() + "\n";
  }
  return output;
}

void CardPile::shuffle() {
  // Fisher yates
  for (auto i = static_cast<int>(m_cards_.size()) - 1; i > 0; i--) {
    const auto j = rand_range(i + 1);
    std::swap(m_cards_[i], m_cards_[j]);
  }
}

File: RustBlackJack/src/cardpile.rs
use crate::card::Card;
use crate::deck::Deck;
use std::time;

pub struct CardPile {
    #[allow(dead_code)]
    pub m_decks: i32,
    pub m_cards: Vec<*mut Card>,
    pub m_original_cards: Vec<*mut Card>,
    pub state: u64,
}

impl CardPile {
    pub fn new(decks: i32) -> CardPile {
        let mut c = Vec::<*mut Card>::with_capacity(decks as usize * 52);
        for _ in 0..decks {
            let mut temp = Deck::new();
            c.append(&mut temp.m_cards);
        }
        let mut cp = CardPile {
            m_decks: decks,
            m_original_cards: c.clone(),
            m_cards: c,
            state: time::SystemTime::now()
                .duration_since(time::SystemTime::UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        };

        cp.refresh();

        cp
    }

    // From https://github.com/lemire/testingRNG
    fn wyrand(&mut self) -> u64 {
        self.state = self.state.wrapping_add(0xa0761d6478bd642f);
        let t: u128 = (self.state as u128).wrapping_mul((self.state ^ 0xe7037ed1a0b428db) as u128);
        (t.wrapping_shr(64) ^ t) as u64
    }

    // use nearly divisionless technique found here https://github.com/lemire/FastShuffleExperiments
    fn rand_range(&mut self, s: u64) -> u64 {
        let mut x = self.wyrand();
        let mut m = x as u128 * s as u128;
        let mut l = m as u64;
        if l < s {
            let thresh = s.wrapping_neg() % s;
            while l < thresh {
                x = self.wyrand();
                m = x as u128 * s as u128;
                l = m as u64;
            }
        }
        (m >> 64) as u64
    }

    // pub fn print(&self) -> String {
    //     self.m_cards[0].print();
    //     let mut output = String::default();
    //     for card in self.m_cards.iter() {
    //         output += card.print();
    //     }
    //     return output;
    // }

    pub fn shuffle(&mut self) {
        for i in (1..self.m_cards.len()).rev() {
            let j = self.rand_range((i + 1) as u64) as usize;
            self.m_cards.swap(i, j);
        }
    }

    pub fn refresh(&mut self) {
        self.m_cards = self.m_original_cards.clone();
    }
}
